Tracing: Migrate to OpenTelemetry library (#9724)
prometheus/prometheus: docs/configuration/configuration.md

# Storage related settings that are runtime reloadable.
storage:
  [ - <exemplars> ... ]

# Tracing configures exporting traces from Prometheus to a tracing
# backend via the OTLP protocol. Tracing is currently an experimental

> # feature and could change in the future.

Author: Amended, thanks!

# Storage related settings that are runtime reloadable.
storage:
  [ - <exemplars> ... ]

# Tracing configures exporting traces from Prometheus to a tracing
# backend via the OTLP protocol. Tracing is currently an experimental
# feature and could change in the future.
tracing:
  # Client used to export the traces. Options are 'http' or 'grpc'.
  client_type: <string>

  # Endpoint to send the traces to.
  endpoint: <string>

  # Sets the probability a given trace will be sampled. Must be a float from 0 through 1.
  [ sampling_fraction: <float> | default = 0 ]

  # If enabled, the client will use a secure connection.

>   [ with_secure: <bool> | default = false ]

Reviewer: Sorry, my request was unclear here. What I meant was a stub like this:

```yaml
tracing:
  [ <tracing> ]
```

I think the config should still have a separate section like before.

Author: Fixed, thanks :raised_hands:

  [ - <relabel_config> ... ]
```

> ### `<tracing>`

Author: Thanks for pointing this out :+1: I adjusted to documentation accordingly.

  [ - <relabel_config> ... ]
```

### `<tracing>`
`tracing` configures exporting traces from Prometheus to a tracing backend via the OTLP protocol.

Must be used together with `--enable-feature=tracing` flag.

```yaml
# Decides whether to use HTTP or gRPC client to export the traces.
[ client_type: <string> | default = grpc ]

# Service name to be used for tracing purposes.
[ service_name: <string> | default = prometheus ]

# Endpoint to send the traces to. By default, this is the Otel Collector's gRPC port 7314 on a localhost.
[ endpoint: <string> | default = localhost:7314 ]

> # Sets the probability a given trace will be sampled. Must be a float between 0 and 1. The default is 0, meaning no traces are exported.

Reviewer: The range is inclusive of 0 and 1 https://pkg.go.dev/go.opentelemetry.io/otel/sdk@v1.2.0/trace#TraceIDRatioBased.

```suggestion
# Sets the probability a given trace will be sampled. Must be a float from 0 through 1. The default is 0, meaning no traces are exported.
```

Reviewer: As pointed out by @roidelapluie, having a sampling fraction of 0 means Prometheus won't create any new traces but still continue/export existing traces from other services.

```suggestion
# Sets the probability a given trace will be sampled. Must be a float between 0 and 1.
```

Author: Agreed, that's a more precise statement on behavior, I'll remove the second sentence.
